### ISSUE TITLE
Recurring `command not found` error in subshells

### ISSUE THREAD

**Issue by claui** [Bash-it/bash-it#708]
## tl;dr

Launching a subshell directly from the OS X Terminal causes recurring `command not found` errors with most themes.
## Prerequisites
- OS X 10.10.5 (but the issue should occur with any OS X version)
- Default system bash (but the issue should occur with any bash version)
- Some fancy theme needs to be active, e. g. `sexy` or `pure`; the issue is reproducible with any theme which defines (but doesn’t bother to export) a dynamic `PROMPT_COMMAND` which calls one or more custom shell functions (e. g. `prompt_command`)
- No `~/.bashrc` present; you keep all your stuff in `~/.bash_profile` like every good OS X citizen does
- Pristine `/etc/profile` (unchanged from factory default)
- Pristine `/etc/bashrc` (unchanged from factory default)
## Steps to reproduce

Some of the themes that ship with Bash-it get fancy with dynamic `PROMPT_COMMAND`s backed by shell functions.

If you have one of those themes active and then type `bash` inside the OS X Terminal, your custom prompt will be lost:

```
(^_^) "`-._,>>-[ 10:09am ]-._,-'(AAPL $127.50)--->$ bash
bash-3.2$ 
```

Since last year, recurring error messages have started to appear along with it:

```
(O_o) "`-._,>>-[ 10:09am ]-._,-'(AAPL $105.97)--->$ bash
bash: prompt_command: command not found
bash-3.2$ echo Y U NO PROMPT
Y U NO PROMPT
bash: prompt_command: command not found
bash-3.2$ 
```
## Analysis

While setting up itself, a Bash theme typically does not bother to export `PROMPT_COMMAND`, or associated private functions it has defined. In practice, it doesn’t have to: in the Linux world, all interactive shells are **non-login shells.** Since Bash-it resides inside `.bashrc`, it will get invoked along with every such shell.  
So whenever an interactive shell process spawns, `.bash_rc` sources Bash-it which sources the theme, which then in turn sets up itself.

If you’re a Mac user you might have learned that OS X Terminal is an entirely different beast which loves to spawn **login shells** left and right when you open new Terminal windows.

Bash-it tries to adapt to this weirdness by hooking itself inside `.bash_profile` instead of `.bashrc`. While this is a clever workaround and works well, it is not perfect; when you type `bash` from inside the OS X Terminal, you’ll find your fancy custom prompt will be lost:

```
(^_^) "`-._,>>-[ 10:09am ]-._,-'(AAPL $127.50)--->$ bash
bash-3.2$ 
```

How did the subshell lose the custom prompt? Like explained above, themes tend to skimp on exporting their prompt fu. Consequently, launching a subshell from within OS X Terminal is not going to inherit the prompt by default. A minor annoyance but kind of works as designed.

Or more precisely, it _used_ to work – until PR #514 introduced code to handle `update_terminal_cwd` on OS X:

```
export PROMPT_COMMAND="$PROMPT_COMMAND;update_terminal_cwd"
```

Now, launching a `bash` subshell from OS X Terminal causes error messages to appear with every prompt:

```
(O_o) "`-._,>>-[ 10:09am ]-._,-'(AAPL $105.97)--->$ bash
bash: prompt_command: command not found
bash-3.2$ echo Y U NO PROMPT
Y U NO PROMPT
bash: prompt_command: command not found
bash-3.2$ 
```

Those `command not found` errors are caused by a regression in PR #514 attempting to inject `update_terminal_cwd` into the prompt, which as a side effect also escalates the exportedness of `PROMPT_COMMAND`. However, the backing shell functions (which are theme-specific, powering the dynamic prompt) remain unexported.

The subshell cannot recover from this partially broken environment because unlike in Linux, Bash-it on OS X dwells in `.bash_profile` but the latter is not sourced by non-login subshells. So Bash-it will not be sourced thus the backing shell functions are not redefined; since those functions were also never exported, the subshell shows an error message every time it wants to eval `PROMPT_COMMAND`.
## Proposed fix

Obviously, it is not Bash-it’s job to anticipate all possible names of fancy shell functions which `PROMPT_COMMAND` might call.

To fix the issue while keeping existing themes working on every platform, I suggest that Bash-it try to preserve the exportedness when tampering with `PROMPT_COMMAND`. It should be up to the individual theme whether it wants to export all the things or nothing.

**PS.** The prompt in the examples is not my real prompt. I just made it up.

**Comment by nwinkler** [comment]
Merged #709 that fixes this error.

**Comment by kbsali** [comment]
@nwinkler i'm having the same issue : 
```
 kevin  ~  130   echo bla
bla
prompt_command: command not found
```
(I wish I could tell you which version I'm using... but I cannot! ;) )

**Comment by nwinkler** [comment]
Which Bash-it theme are you using?

**Comment by kbsali** [comment]
powerline-plain,
but now i'm seeing that they issue is only when i'm in a byobu/tmux session...
